src/FFNetAgent.cpp

```diff
@@ -0,0 +1,362 @@
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::min_freq_signal_name = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MIN_AVAIL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MIN_AVAIL"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::max_freq_control_name = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MAX_CONTROL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MAX_CONTROL"}
+            };
+    const std::map<geopm_domain_e, std::string> FFNetAgent::min_freq_control_name = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "CPU_FREQUENCY_MIN_CONTROL"},
+                {GEOPM_DOMAIN_GPU, "GPU_CORE_FREQUENCY_MIN_CONTROL"}
+            };
+
+    const std::map<geopm_domain_e, std::string> FFNetAgent::trace_suffix = 
+            {
+                {GEOPM_DOMAIN_PACKAGE, "_cpu_"},
+                {GEOPM_DOMAIN_GPU, "_gpu_"}
+            };
+
+    FFNetAgent::FFNetAgent()
+        : FFNetAgent(platform_io(), platform_topo())
+    {
+
+    }
+
+    FFNetAgent::FFNetAgent(
+            PlatformIO &plat_io,
+            const PlatformTopo &topo,
+            std::map<std::pair<geopm_domain_e, int>, std::shared_ptr<DomainNetMap> > &net_map,
+            std::map<geopm_domain_e, std::shared_ptr<RegionHintRecommender> > &freq_recommender
+            )
+        : m_platform_io(plat_io)
+          , m_last_wait{{0, 0}}
+          , M_WAIT_SEC(0.020) // 20ms Wait
+          , m_perf_energy_bias(0)
+    {
+        geopm_time(&m_last_wait);
+        init_domain_indices(topo);
+
+        for (const geopm_domain_e domain_type : m_domain_types) {
+            m_freq_recommender[domain_type] = freq_recommender.at(
+                        domain_type
+                    );
+        }
+
+        for (const m_domain_key_s domain_key : m_domains) {
+            m_net_map[domain_key] = net_map.at(
+                    std::make_pair(
+                        domain_key.type,
+                        domain_key.index)
+                    );
+        }
+    }
+
+    FFNetAgent::FFNetAgent(PlatformIO &plat_io, const PlatformTopo &topo)
+        : m_platform_io(plat_io)
+          , m_last_wait{{0, 0}}
+          , M_WAIT_SEC(0.050) // 50ms Wait
+          , m_perf_energy_bias(0)
+    {
+        geopm_time(&m_last_wait);
+        init_domain_indices(topo);
+
+        //Loading neural nets
+        for (const m_domain_key_s domain_key : m_domains) {
+            m_net_map[domain_key] = std::make_shared<DomainNetMapImp>(
+                            getenv(nnet_envname.at(domain_key.type)),
+                            domain_key.type,
+                            domain_key.index);
+        }
+
+        for (geopm_domain_e domain_type : m_domain_types) {
+            m_freq_recommender[domain_type] = std::make_shared<RegionHintRecommenderImp>(
+                    getenv(freqmap_envname.at(domain_type)),
```

Comment:
Calling getenv() is generally to be avoided if possible.  Why is more than one environment variable needed if there is a requirement?  Is this something that could be managed by the Environment class?